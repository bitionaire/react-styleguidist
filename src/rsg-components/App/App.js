import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { isFunction } from 'lodash';
import slots from 'rsg-components/slots';
import StyleGuide from 'rsg-components/StyleGuide';
import getPageTitle from '../../utils/getPageTitle';
import getRouteData from '../../utils/getRouteData';
import globalizeComponents from '../../utils/globalizeComponents';
import processSections from '../../utils/processSections';

export default class App extends Component {
	static propTypes = {
		/** Object returned by styleguide-loader */
		styleguide: PropTypes.object.isRequired,
		codeRevision: PropTypes.number.isRequired,
		loc: PropTypes.object,
		doc: PropTypes.object,
		hist: PropTypes.object,
	};

	static defaultProps = {
		loc: window.location,
		doc: document,
		hist: window.history,
	};

	constructor(props) {
		super(props);
		this.lastHash = props.loc.hash;
	}

	componentDidUpdate() {
		// scroll to top of styleguide container when sections changed
		if (this.props.styleguide.config.pagePerSection && this.props.loc.hash !== this.lastHash) {
			if (window && isFunction(window.scrollTo)) {
				window.scrollTo(0, 0);
			}
		}
		this.currentHash = this.props.loc.hash;
	}

	render() {
		const { styleguide, codeRevision, loc, doc, hist } = this.props;

		const allSections = processSections(styleguide.sections);

		// Globalize all components, not just ones we see on the screen, to make
		// all components accessible to all examples
		globalizeComponents(allSections);

		const { title, pagePerSection } = styleguide.config;
		const { sections, displayMode } = getRouteData(allSections, loc.hash, pagePerSection);

		// Update page title
		doc.title = getPageTitle(sections, title, displayMode);

		// If the current hash location was set to just `/` (e.g. when navigating back from isolated view to overview)
		// replace the URL with one without hash, to present the user with a single address of the overview screen
		if (loc.hash === '#/') {
			const url = loc.pathname + loc.search;
			hist.replaceState('', doc.title, url);
		}

		return (
			<StyleGuide
				codeRevision={codeRevision}
				config={styleguide.config}
				slots={slots}
				welcomeScreen={styleguide.welcomeScreen}
				patterns={styleguide.patterns}
				sections={sections}
				allSections={allSections}
				displayMode={displayMode}
				pagePerSection={pagePerSection}
			/>
		);
	}
}
