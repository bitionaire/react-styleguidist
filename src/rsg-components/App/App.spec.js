import React from 'react';
import { cloneDeep } from 'lodash';
import App from './App';

const styleguide = {
	config: {
		title: 'My Style Guide',
	},
	sections: [
		{
			components: [
				{
					props: {
						displayName: 'Button',
					},
					module: 'ButtonModule',
				},
				{
					props: {
						displayName: 'Image',
					},
					module: 'ImageModule',
				},
			],
		},
	],
	welcomeScreen: false,
	patterns: ['button', 'input'],
};
const codeRevision = 1;
const location = {
	hash: '',
};
const doc = {
	title: () => {},
};
const history = {
	replaceState: () => {},
};

afterEach(() => {
	delete global.Button;
	delete global.Image;
});

describe('<App />', () => {
	it('should render StyleGuide component', () => {
		const result = shallow(
			<App
				styleguide={styleguide}
				codeRevision={codeRevision}
				loc={location}
				doc={doc}
				hist={history}
			/>
		);
		expect(result).toMatchSnapshot();
	});

	it('should globalize all components', () => {
		shallow(
			<App
				styleguide={styleguide}
				codeRevision={codeRevision}
				loc={location}
				doc={doc}
				hist={history}
			/>
		);
		expect(global.Button).toBe('ButtonModule');
		expect(global.Image).toBe('ImageModule');
	});

	it('should globalize all components in isolated mode', () => {
		shallow(
			<App
				styleguide={styleguide}
				codeRevision={codeRevision}
				loc={{ hash: '#!/Button' }}
				doc={doc}
				hist={history}
			/>
		);
		expect(global.Button).toBe('ButtonModule');
		expect(global.Image).toBe('ImageModule');
	});

	it('should change document title', () => {
		const title = jest.fn();
		const location = { hash: '' };
		Object.defineProperty(location, 'title', {
			set: title,
		});
		shallow(
			<App
				styleguide={styleguide}
				codeRevision={codeRevision}
				loc={location}
				doc={location}
				hist={history}
			/>
		);
		expect(title).toBeCalledWith('My Style Guide');
	});

	it('should change document title in isolated mode', () => {
		const title = jest.fn();
		const location = { hash: '#!/Button' };
		Object.defineProperty(location, 'title', {
			set: title,
		});
		shallow(
			<App
				styleguide={styleguide}
				codeRevision={codeRevision}
				loc={location}
				doc={location}
				hist={history}
			/>
		);
		expect(title).toBeCalledWith('Button â€” My Style Guide');
	});

	it('should remove #/ from the address bar', () => {
		const location = { hash: '#/', pathname: '/pizza', search: '?foo=bar' };
		const history = { replaceState: jest.fn() };
		shallow(
			<App
				styleguide={styleguide}
				codeRevision={codeRevision}
				loc={location}
				doc={doc}
				hist={history}
			/>
		);
		expect(history.replaceState).toBeCalledWith('', 'My Style Guide', '/pizza?foo=bar');
	});

	it('should scroll to top on sections change, when pagePerSections is true', () => {
		const scrollTo = jest.fn();
		global.window.scrollTo = scrollTo;

		const styleguideCopy = cloneDeep(styleguide);
		styleguideCopy.config.pagePerSection = true;

		const props = {
			styleguide: styleguideCopy,
			codeRevision,
			loc: { hash: '#A' },
			doc,
			hist: history,
		};

		const wrapper = shallow(<App {...props} />);
		wrapper.setProps({ ...props, loc: { hash: '#B' } });

		expect(scrollTo).toHaveBeenCalled();
	});

	it('should not scroll to top on sections change, when pagePerSections is false', () => {
		const scrollTo = jest.fn();
		global.window.scrollTo = scrollTo;

		const styleguideCopy = cloneDeep(styleguide);
		styleguideCopy.config.pagePerSection = false;

		const props = {
			styleguide: styleguideCopy,
			codeRevision,
			loc: { hash: '#A' },
			doc,
			hist: history,
		};

		const wrapper = shallow(<App {...props} />);
		wrapper.setProps({ ...props, loc: { hash: '#B' } });

		expect(scrollTo).not.toHaveBeenCalled();
	});
});
