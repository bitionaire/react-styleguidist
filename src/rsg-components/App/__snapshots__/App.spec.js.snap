// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render StyleGuide component 1`] = `
<StyleGuide
  allSections={
    Array [
      Object {
        "components": Array [
          Object {
            "module": "ButtonModule",
            "name": "Button",
            "props": Object {
              "displayName": "Button",
              "examples": Array [],
            },
          },
          Object {
            "module": "ImageModule",
            "name": "Image",
            "props": Object {
              "displayName": "Image",
              "examples": Array [],
            },
          },
        ],
        "sections": Array [],
      },
    ]
  }
  codeRevision={1}
  config={
    Object {
      "title": "My Style Guide",
    }
  }
  displayMode="all"
  patterns={
    Array [
      "button",
      "input",
    ]
  }
  sections={
    Array [
      Object {
        "components": Array [
          Object {
            "module": "ButtonModule",
            "name": "Button",
            "props": Object {
              "displayName": "Button",
              "examples": Array [],
            },
          },
          Object {
            "module": "ImageModule",
            "name": "Image",
            "props": Object {
              "displayName": "Image",
              "examples": Array [],
            },
          },
        ],
        "sections": Array [],
      },
    ]
  }
  slots={
    Object {
      "componentToolbar": Array [
        [Function],
      ],
      "docsTabButtons": Array [
        Object {
          "id": "rsg-usage",
          "render": [Function],
        },
      ],
      "docsTabs": Array [
        Object {
          "id": "rsg-usage",
          "render": [Function],
        },
      ],
      "exampleTabButtons": Array [
        Object {
          "id": "rsg-code-editor",
          "render": [Function],
        },
      ],
      "exampleTabs": Array [
        Object {
          "id": "rsg-code-editor",
          "render": [Function],
        },
      ],
      "exampleToolbar": Array [
        [Function],
      ],
      "sectionToolbar": Array [
        [Function],
      ],
    }
  }
  welcomeScreen={false}
/>
`;
